src/k2/cpo/service/Service.cpp

```diff
@@ -147,9 +163,86 @@ seastar::future<> CPOService::start() {
         }
     }
 
-    return seastar::make_ready_future<>();
+    return AppBase().getDist<HealthMonitor>().local().getTSOEndpoints()
+    .then([this] (std::vector<String>&& eps) {
+        size_t tsoID{1000};
+        for (const auto & ep : eps) {
+            _failedTSOs[ep] = tsoID++;
+        }
+        _tsoAssignTimer.setCallback([this] {
+            return _assignAllTSOs()
+            .then([this] (bool success) {
+                if (!success) {
+                    K2LOG_W(log::cposvr, "Failed to assign all TSOs");
+                    _tsoAssignTimer.arm(_assignTimeout());
+                }
+            });
+        });
+        _tsoAssignTimer.arm(0s);
+    });
+
+}
+
+seastar::future<bool> CPOService::_assignAllTSOs() {
+    std::vector<seastar::future<>> futs;
+    for (auto tso=_failedTSOs.begin(); tso!=_failedTSOs.end(); ++tso) {
+        futs.push_back(
+            _assignTSO(tso->first, tso->second, _maxAssignRetries())
+        );
+    }
+    return seastar::when_all_succeed(futs.begin(), futs.end()).discard_result()
+    .then([this] {
+        return _failedTSOs.empty();
+    });
+}
+
+seastar::future<> CPOService::_doAssignTSO(const String &ep, size_t tsoID) {
+    k2::dto::AssignTSORequest request{.tsoID = tsoID, .tsoErrBound=_TSOErrorBound()};
+    auto txep = RPC().getTXEndpoint(ep);
+    K2LOG_D(log::cposvr, "assigning TSO: {}", ep);
+    if (!txep) {
+        K2LOG_W(log::cposvr, "unable to obtain endpoint for {}", ep);
+        return seastar::make_exception_future(StopRetryException{});
+    }
+    return RPC().callRPC<dto::AssignTSORequest, dto::AssignTSOResponse>(
+        dto::Verbs::TSO_ASSIGNMENT, request, *txep, _assignTimeout()
+    ).then([&ep,tsoID,this] (auto &&result) {
```

Comment:
I don't think it is safe to capture ep by reference here, you can capture by value instead.